lib/gem_content/package.rb

```diff
@@ -0,0 +1,31 @@
+# A wrapper around Gem::Package to add behaviors we need for working with
+# individual files in memory.
+class GemContent::Package
+  attr_reader :package
+
+  def initialize(gem)
+    @package = gem.is_a?(::Gem::Package) ? gem : ::Gem::Package.new(gem)
+  end
+
+  delegate :verify, :spec, :contents, to: :package
+
+  def each(&)
+    open_data_tar do |data_tar|
+      data_tar.each do |entry|
+        begin
+          yield GemContent::Entry.from_package_entry(entry)
+        rescue StandardError => e
+          Rails.error.report(e, handled: true)
+        end
+      end
+    end
+  end
+
+  def open_data_tar(&)
+    package.gem.with_read_io do |io|
+      Gem::Package::TarReader.new(io).seek("data.tar.gz") do |gem_entry|
```

Comment:
do we also want to be saving the gemspec & whatever the other tar in a `.gem` is?

Reply:
I save the spec from Gem::Package#spec.to_ruby. The other file is the checksum that verifies the other two files are not corrupt.